Service/Product/MissingPriceIndexHandler.php

```diff
@@ -0,0 +1,193 @@
+<?php
+
+namespace Algolia\AlgoliaSearch\Service\Product;
+
+use Algolia\AlgoliaSearch\Helper\Logger;
+use Magento\Catalog\Model\ResourceModel\Product\Collection as ProductCollection;
+use Magento\Catalog\Model\ResourceModel\Product\CollectionFactory;
+use Magento\Framework\App\ResourceConnection;
+use Magento\Framework\DB\Select;
+use Magento\Framework\Indexer\IndexerInterface;
+use Magento\Framework\Indexer\IndexerRegistry;
+use Magento\Framework\Indexer\StateInterface;
+use Zend_Db_Select;
+
+class MissingPriceIndexHandler
+{
+    public const PRICE_INDEX_TABLE = 'catalog_product_index_price';
+    public const PRICE_INDEX_TABLE_ALIAS = 'price_index';
+    public const MAIN_TABLE_ALIAS = 'e';
+
+    protected array $_indexedProducts = [];
+
+    protected IndexerInterface $indexer;
+    public function __construct(
+        protected CollectionFactory $productCollectionFactory,
+        protected ResourceConnection $resourceConnection,
+        protected Logger $logger,
+        IndexerRegistry $indexerRegistry
+    )
+    {
+        $this->indexer = $indexerRegistry->get('catalog_product_price');
+    }
+
+    /**
+     * @param string[]|ProductCollection $products
+     * @return string[] Array of product IDs that were reindexed by this repair operation
+     * @throws \Zend_Db_Select_Exception
+     */
+    public function refreshPriceIndex(array|ProductCollection $products): array
+    {
+        $reindexIds = $this->getProductIdsToReindex($products);
+        if (empty($reindexIds)) {
+            return [];
+        }
+
+        $this->logger->log(__("Pricing records missing or invalid for %1 product(s)", count($reindexIds)));
+        $this->logger->log(__("Reindexing product ID(s): %1", implode(', ', $reindexIds)));
+
+        $this->indexer->reindexList($reindexIds);
+
+        return $reindexIds;
+    }
+
+    /**
+     * Analyzes a product collection and determines which (if any) records should have their prices reindexed
+     * @param string[]|ProductCollection $products - either an explicit list of product ids or a product collection
```

Comment:
Is there already a case where an array of product ids is passed instead of a ProductCollection ?

Reply:
I wanted to keep it flexible. I had an earlier iteration where I was passing IDs but all jobs seem to converge at a single point namely `\Algolia\AlgoliaSearch\Helper\Data::rebuildStoreProductIndexPage`. There are problems with the implementation of that function and we may want to revisit. Also the extra collection operations (which are necessary to keep this code minimally invasive on the regular operation of the query) increase processing time marginally so their may be opportunity for future optimizations where we don't need to do this.